[ONNX] Extract export verification as standalone api from unittest

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #78323
* #77289
* __->__ #76929

The verification logic is refactored and extracted from
`test_pytorch_onnx_onnxruntime.py`.

BowenBao committed:
[ONNX] Extract export verification as standalone api from unittest

The verification logic is refactored and extracted from
`test_pytorch_onnx_onnxruntime.py`.

[ghstack-poisoned]

justinchuby commented on torch/onnx/verify.py:
```diff
@@ -0,0 +1,262 @@
+import copy
```
I would add a summary on the top of the file for this module.

justinchuby commented on torch/onnx/verify.py:
```diff
@@ -0,0 +1,262 @@
+import io
+import os
+import tempfile
+from contextlib import ExitStack
```
I would avoid importing a class from a module. Import the module instead to keep the name space clean and allow readers more context on where the class comes from (below when being used). https://google.github.io/styleguide/pyguide.html#22-imports

justinchuby commented on torch/onnx/verify.py:
```diff
@@ -0,0 +1,262 @@
+
+def _flatten_tuples(elem):
+    flattened = []
+    for t in elem:
```
Avoid names that are too short: https://google.github.io/styleguide/pyguide.html#3161-names-to-avoid

BowenBao replied:
feels alright falling under
```
single character names, except for specifically allowed cases:

* counters or iterators (e.g. i, j, k, v, et al.)
```

justinchuby replied:
ok

justinchuby commented on torch/onnx/verify.py:
```diff
@@ -0,0 +1,262 @@
+
+
+def _inline_flatten_list(inputs, res_list):
+    for i in inputs:
```
naming: ditto

BowenBao replied:
ditto

justinchuby replied:
`i` should specifically be a counter, taking numbers as values and most likely be an integer. Here i stands for input, which imo should be spelled out. Very nick picky though. I would do

```
for input_ in inputs:
    ...
```

or `val`, or `elem`.

justinchuby commented on torch/onnx/verify.py:
```diff
@@ -0,0 +1,262 @@
+    opset_version: Optional[int] = None,
+    keep_initializers_as_inputs: bool = True,
+):
+    """Verify model export to ONNX with ONNXRuntime."""
```
I would add docstrings and type annotations for the public functions.

justinchuby commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -929,9 +653,7 @@ def test_mask_rcnn(self):
-        from torchvision.models.detection.roi_heads import (
-            heatmaps_to_keypoints,
-        )
+        from torchvision.models.detection.roi_heads import heatmaps_to_keypoints
```
I would move this out to the top to make the dependency clear.

BowenBao replied:
+1

justinchuby commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -584,7 +308,7 @@ def forward(self, input):
 
         x = torch.tensor([2], dtype=torch.long)
-        self.run_model_test_with_external_data(LargeModel(), x)
+        run_model_test_with_external_data(self, LargeModel(), x)
```
Love this refactoring - makes the teat class clean

justinchuby reviewed (commented):
Some initial comments. Will take a closer look tomorrow.

justinchuby commented:
I would hide the verify module before it is ready to be exposed.

justinchuby commented:
I suggest using more descriptive names for modules and action verbs for functions. Examples are `torch.special`, `torch.distributed`, `torch.testing` etc.

What do you think of `verify` -> `verification`?

justinchuby commented on torch/onnx/verify.py:
```diff
@@ -0,0 +1,262 @@
+                tmpdirname = stack.enter_context(tempfile.TemporaryDirectory())
+                model_f = os.path.join(tmpdirname, "model.onnx")
+
+            ort_sess = _convert_to_onnx(
```
naming: abbr

BowenBao replied:
ort_session seems reasonable

justinchuby commented on torch/onnx/verify.py:
```diff
@@ -0,0 +1,262 @@
+    verbose=False,
+    ort_providers=_ORT_PROVIDERS,
+    ort_optim_on=True,
+    use_external_data=False,
```
I am trying to understand what the behavior is when use_external_data is True. Does it output an onnx model to the temp dir and then verifying it? I am thinking how best the option can be exposed.

BowenBao replied:
yes. The underlying restriction is that dir path is required for external export, where as io.BytesIO suffices otherwise. This detail is hidden away by flag `use_external_data`.

justinchuby commented on torch/onnx/verify.py:
```diff
@@ -0,0 +1,262 @@
+    atol=1e-7,
+    do_constant_folding=True,
+    dynamic_axes=None,
+    test_with_inputs=None,
```
`additional_test_inputs` etc. for clarity? This name feels like a bool param.

BowenBao replied:
+1

justinchuby commented on torch/onnx/verify.py:
```diff
@@ -0,0 +1,262 @@
+    input_names=None,
+    output_names=None,
+    fixed_batch_size=False,
+    dict_check=True,
```
naming: It is unclear to me what "dict" is refering to. Is it checking the input arguments? Should we consider a different name?

BowenBao replied:
+1

justinchuby commented on torch/onnx/verify.py:
```diff
@@ -0,0 +1,262 @@
+    training=None,
+    verbose=False,
+    ort_providers=_ORT_PROVIDERS,
+    ort_optim_on=True,
```
naming: abbr

BowenBao replied:
optim seems a reasonable abbreviation for optimization.

BowenBao replied:
well, optim is not something new, already used by other packages as well. I don't have any strong preference though, just to bring up.

https://pytorch.org/docs/stable/optim.html
https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/optim
> As a rule of thumb, an abbreviation is probably OK if it's listed in Wikipedia.

justinchuby commented on torch/onnx/verify.py:
```diff
@@ -0,0 +1,262 @@
+    fixed_batch_size=False,
+    training=None,
+    verbose=False,
+    ort_providers=_ORT_PROVIDERS,
```
_ORT_PROVIDERS is a list which should not be a default value. Should we change _ORT_PROVIDERS to a tuple? https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

BowenBao replied:
hmm I feel this is acceptable, given that `_ORT_PROVIDERS` is named as constant. And ort_providers is not expected to be mutated. I don't think changing to tuple, or changing to None and setting `ort_providers` inside the function would help very much against unlikely adversarial usage while increasing complexity.

justinchuby replied:
I would worry less about adversarial usage than general good practices and what the code tells the readers. If we pass a list as a default, typical readers will wonder if there is anything special about this argument, and whether `_ORT_PROVIDERS` is expected to be mutated. As far as I can tell it is immutable. It is good to just make it a tuple to (1) clarify the intention and (2) avoid the anti-pattern of supplying a mutable as default

justinchuby commented on torch/onnx/verify.py:
```diff
@@ -0,0 +1,262 @@
+    return elem
+
+
+def _convert_to_onnx(
```
I recommend type annotating this function. It is unclear what some of the params defaulted to None should take.

justinchuby commented on torch/onnx/verify.py:
```diff
@@ -0,0 +1,262 @@
+    return [_to_numpy(v) for v in value_unpacked]
+
+
+def _run_ort(ort_sess, inputs):
```
naming: abbr. Here and elsewhere

justinchuby commented on torch/onnx/verify.py:
```diff
@@ -0,0 +1,262 @@
+    for i, input in enumerate(inputs):
+        if i == len(ort_sess_inputs) or ort_sess_inputs[i].name in ort_inputs:
+            raise ValueError(
+                f"got too many positional inputs. inputs: {inputs}. kw_inputs: {kw_inputs}"
```
I would add (who) got too many positional inputs for clarity.

BowenBao replied:
error will be thrown from _run_ort which should be straight forward, I guess it wouldn't hurt adding that though

justinchuby commented on torch/onnx/verify.py:
```diff
@@ -0,0 +1,262 @@
+
+            # if additional test inputs are provided run the onnx
+            # model with these inputs and check the outputs
+            if test_with_inputs is not None:
```
This function is a little too big. Would it be possible to factor some parts out? For example, the part that prepares the input (247-260) seems to be a good candidate.

BowenBao committed:
Update on "[ONNX] Extract export verification as standalone api from unittest"


The verification logic is refactored and extracted from
`test_pytorch_onnx_onnxruntime.py`.

[ghstack-poisoned]

BowenBao commented on torch/onnx/verify.py:
```diff
@@ -0,0 +1,262 @@
+    elif isinstance(elem, int):
+        return np.array(elem, dtype=int)
+    elif isinstance(elem, float):
+        return np.array(elem, dtype=float)
```
possibly
```python
elif isinstance(elem, (bool, int, float)):
    return np.array(elem, dtype=type(elem))  # or `return np.array(elem)`
```
not sure if why it was the way it was.

BowenBao commented on torch/onnx/verify.py:
```diff
@@ -0,0 +1,262 @@
+    [
+        np.testing.assert_allclose(out, ort_out, rtol=rtol, atol=atol)
+        for out, ort_out in zip(outputs, ort_outs)
+    ]
```
The complexity looks fine under https://google.github.io/styleguide/pyguide.html#27-comprehensions--generator-expressions

justinchuby replied:
I mean, the resulting list is discarded/not used?

BowenBao commented on torch/onnx/verify.py:
```diff
@@ -0,0 +1,262 @@
+    return ort_sess
+
+
+def _inline_flatten_list(inputs, res_list):
```
It appears intended to treat list differently.

BowenBao commented on torch/onnx/verify.py:
```diff
@@ -0,0 +1,262 @@
+        return np.array(elem, dtype=float)
+    elif isinstance(elem, dict):
+        flattened = []
+        for k in elem:
```
Good point, I have no idea what is the logic behind dictionary handling here.

BowenBao commented on torch/onnx/verify.py:
```diff
@@ -0,0 +1,262 @@
+    return flattened
+
+
+def _to_numpy(elem):
```
seems `dict` usage strikes again. My guess is that string is kept as is, some how.

BowenBao commented:
Thanks @justinchuby for thorough review! Many great refactor suggestions. I will create a commit on top of this one to keep this PR mostly moving, and the next one refactoring.

BowenBao committed:
Update on "[ONNX] Extract export verification as standalone api from unittest"


The verification logic is refactored and extracted from
`test_pytorch_onnx_onnxruntime.py`.

[ghstack-poisoned]

BowenBao commented:
> Should we hide the verify module by prepending an underscore to its name before it is ready to be exposed? This way we have more flexibility to refine its api before people starting to rely on it

This is interesting idea. I agree the current convention does not align with standard guide. Current convention is using `__init__.py` to define all public facing apis.

justinchuby reviewed (approved):
LGTM w/ follow up PR on style.

BowenBao committed:
Update on "[ONNX] Extract export verification as standalone api from unittest"


The verification logic is refactored and extracted from
`test_pytorch_onnx_onnxruntime.py`.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Extract export verification as standalone api from unittest"


The verification logic is refactored and extracted from
`test_pytorch_onnx_onnxruntime.py`.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Extract export verification as standalone api from unittest"


The verification logic is refactored and extracted from
`test_pytorch_onnx_onnxruntime.py`.

[ghstack-poisoned]

justinchuby commented on torch/onnx/verification.py:
```diff
@@ -0,0 +1,271 @@
+    elif training is None or training == torch.onnx.TrainingMode.EVAL:
+        model.eval()
+    if input is None:
+        input = torch.randn(batch_size, 3, 224, 224, requires_grad=True)
```
Is the assumption on the input size a reasonable / general enough one?

BowenBao replied:
bc legacy code for caffe2 vision related model tests. I'll comment.

BowenBao replied:
removed in follow up

BowenBao committed:
Update on "[ONNX] Extract export verification as standalone api from unittest"


The verification logic is refactored and extracted from
`test_pytorch_onnx_onnxruntime.py`.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Extract export verification as standalone api from unittest"


The verification logic is refactored and extracted from
`test_pytorch_onnx_onnxruntime.py`.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Extract export verification as standalone api from unittest"


The verification logic is refactored and extracted from
`test_pytorch_onnx_onnxruntime.py`.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Extract export verification as standalone api from unittest"


The verification logic is refactored and extracted from
`test_pytorch_onnx_onnxruntime.py`.

[ghstack-poisoned]

BowenBao commented:
@pytorchbot merge this

BowenBao commented:
cc @garymm  for merge approval

garymm commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -57,228 +57,20 @@
-        ort_inputs[ort_sess_inputs[i].name] = input
-    ort_outs = ort_sess.run(None, ort_inputs)
-    return inline_flatten_list(ort_outs, [])
+def run_model_test(self, *args, **kwargs):
```
I think naming this `self` and not having a type annotation and having it live outside of a class is a bit confusing.
How's about we either move it inside `_TestONNXRuntime` or:
1. Add a comment documenting why we can't
2. Rename self and / or add a type annotation so it's clear what it is.

BowenBao replied:
renamed to `test_suite: _TestONNXRuntime`

garymm commented on torch/onnx/verification.py:
```diff
@@ -0,0 +1,271 @@
+"""Functions to verify exported ONNX model against original PyTorch model.
+
+ONNXRuntime is required, and is used as the ONNX backend for export verification.
+The goal is to verify that the exported ONNX model is functionally equivalent to the
```
I'd remove this sentence. It's pretty much the same as the sentence on line 1.

garymm commented on torch/onnx/verification.py:
```diff
@@ -0,0 +1,271 @@
+"""Functions to verify exported ONNX model against original PyTorch model.
+
+ONNXRuntime is required, and is used as the ONNX backend for export verification.
```
```suggestion
ONNX Runtime is required, and is used as the ONNX backend for export verification.
```

garymm reviewed (approved):
Please address minor feedback and then feel free to merge.

BowenBao committed:
Update on "[ONNX] Extract export verification as standalone api from unittest"


The verification logic is refactored and extracted from
`test_pytorch_onnx_onnxruntime.py`.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Extract export verification as standalone api from unittest"


The verification logic is refactored and extracted from
`test_pytorch_onnx_onnxruntime.py`.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Extract export verification as standalone api from unittest"


The verification logic is refactored and extracted from
`test_pytorch_onnx_onnxruntime.py`.

[ghstack-poisoned]

BowenBao commented:
@pytorchbot merge this
